tests/integration_tests/functional/test_feat_parity.py

```diff
@@ -0,0 +1,155 @@
+    lst3 = [value for value in lst1 if value in lst2]
+    return lst3
+
+
+def get_guest_kernel_ver(vm):
+    """
+    Return the guest kernel version.
+    Useful when running test matrix with multiple guest kernels.
+    """
+    ssh_conn = net_tools.SSHConnection(vm.ssh_config)
+    read_kernel_ver_cmd = "uname -r"
+    _, stdout, stderr = ssh_conn.execute_command(read_kernel_ver_cmd)
+    assert stderr.read() == ""
+    return stdout.read().strip()
+
+
+def _test_cpuid_feat_flags(context):
+    vm_builder = context.custom["builder"]
+    root_disk = context.disk.copy()
+    cpu_template = context.custom["cpu_template"]
+    must_be_set = context.custom["flags_must_be_set"]
+    must_be_unset = context.custom["flags_must_be_unset"]
+
+    vm_instance = vm_builder.build(
+        kernel=context.kernel,
+        disks=[root_disk],
+        ssh_key=context.disk.ssh_key(),
+        config=context.microvm,
+        cpu_template=cpu_template,
+    )
+    vm = vm_instance.vm
+    vm.start()
+
+    cpuid = cpuid_utils.get_guest_cpuid(vm)
+    kernel_ver = get_guest_kernel_ver(vm)
+    allowed_regs = ["eax", "ebx", "ecx", "edx"]
+
+    for leaf, subleaf, reg, flags in must_be_set:
+        assert reg in allowed_regs
+        actual = cpuid[(leaf, subleaf, reg)] & flags
+        expected = flags
+        assert (
+            actual == expected
+        ), f"{cpu_template}: {kernel_ver=} {leaf=:#x} {subleaf=:#x} {reg=} {actual=:#x}, {expected=:#x}"
+
+    for leaf, subleaf, reg, flags in must_be_unset:
+        assert reg in allowed_regs
+        actual = cpuid[(leaf, subleaf, reg)] & flags
+        expected = 0
+        assert (
+            actual == expected
+        ), f"{cpu_template} {kernel_ver=} {leaf=:#x} {subleaf=:#x} {reg=} {actual=:#x}, {expected=:#x}"
+
+
+def _test_cpuid_feat_flags_matrix(
+    bin_cloner_path,
+    network_config,
+    cpu_template,
+    flags_must_be_set,
+    flags_must_be_unset,
+):
+    """
+    This launches tests matrix for CPUID feature flag checks for the given CPU template.
+    """
+
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    # Testing matrix:
+    # - Guest kernel: Linux 4.14 & Linux 5.10
+    # - Rootfs: Ubuntu 18.04 with msr-tools package installed
+    # - Microvm: 1vCPU with 1024 MB RAM
+    microvm_artifacts = ArtifactSet(artifacts.microvms(keyword="1vcpu_1024mb"))
+    kernel_artifacts = ArtifactSet(artifacts.kernels())
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword="bionic-msrtools"))
+    assert len(disk_artifacts) == 1
+
+    test_context = TestContext()
+    test_context.custom = {
+        "builder": MicrovmBuilder(bin_cloner_path),
+        "network_config": network_config,
+        "cpu_template": cpu_template,
+        "flags_must_be_set": flags_must_be_set,
+        "flags_must_be_unset": flags_must_be_unset,
+    }
+    test_matrix = TestMatrix(
+        context=test_context,
+        artifact_sets=[microvm_artifacts, kernel_artifacts, disk_artifacts],
+    )
```

Comment:
See https://github.com/firecracker-microvm/firecracker/commit/b646a3388c8ee5d89225e399c78d0571210512c0 for an example of of using fixtures instead of TestMatrix.

Reply:
If I do that, I would need to repeat the following for each test:
```python
@pytest.mark.parametrize(
    "disk", ARTIFACTS.disks(keyword="bionic-msrtools"), ids=lambda disk: disk.name()
)
@pytest.mark.parametrize(
    "kernel", ARTIFACTS.kernels(), ids=lambda kernel: kernel.name()
)
@pytest.mark.parametrize(
    "microvm", ARTIFACTS.microvms(keyword="1vcpu_1024mb"), ids=lambda uvm: uvm.name()
)
@pytest.mark.parametrize(
    "cpu_template",
    cputmpl_utils.select_supported_cpu_templates(cputmpl_utils.ALL_TEMPLATES),
)
def test_feat_parity_cpuid_mpx(bin_cloner_path, network_config, microvm, kernel, disk, cpu_template):
```
Do you think there is a way to avoid that?

Reply:
I think we could re-use some common fixtures in conftest and add a few specific ones for this file. I haven't tested but I think something like this should work:

```python
# Define common fixtures

@pytest.fixture(
    params=ARTIFACTS.microvms(keyword="1vcpu_1024mb"), ids=lambda uvm: uvm.name()
)
def microvm(request):
    """Common fixture for tests in this file"""
    return request.param

@pytest.fixture(
    params=cputmpl_utils.select_supported_cpu_templates(cputmpl_utils.ALL_TEMPLATES),
)
def cpu_template(request):
    return request.param

# Then each test becomes

@pytest.mark.parametrize("rootfs", params=ARTIFACTS.disks("bionic-msrtools"), indirect=True)
def test_feat_parity_cpuid_mpx(bin_cloner_path, network_config, microvm, guest_kernel, rootfs, cpu_template):
    pass
```

Reply:
done